numpy/linalg/tests/test_linalg.py

```diff
@@ -845,6 +846,111 @@ def test_empty(self):
         assert_equal(norm(array([], dtype=self.dt)), 0.0)
         assert_equal(norm(atleast_2d(array([], dtype=self.dt))), 0.0)
 
+    def test_vector_types(self):
+        a = np.array([1, 0, 1])
+
+        exact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllInteger']
+        ]
+
+        inexact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllFloat']
+        ]
+
+        all_types = exact_types + inexact_types
+
+        for each_inexact_types in all_types:
+            at = a.astype(each_inexact_types)
+
+            n = norm(at, -np.inf)
+            assert_(issubclass(n.dtype.type, np.inexact))
+            assert_almost_equal(n, 0.0)
+
+            with warnings.catch_warnings() as w:
+                warnings.simplefilter("ignore", RuntimeWarning)
+                n = norm(at, -1)
+                assert_(issubclass(n.dtype.type, np.inexact))
+                assert_almost_equal(n, 0.0)
+
+            n = norm(at, 0)
+            # Trying to assert equality to `np.dtype(int).type` fails on
+            # 32-bit platforms as it still becomes `np.int64` instead of
+            # `np.int32`. So, this is our workaround.
+            assert_(n.dtype.type in [np.int32, np.int64])
+            assert_almost_equal(n, 2)
+
+            n = norm(at, 1)
+            assert_(issubclass(n.dtype.type, np.inexact))
+            assert_almost_equal(n, 2.0)
+
+            n = norm(at, 2)
+            assert_(issubclass(n.dtype.type, np.inexact))
+            assert_almost_equal(n, 2.0**(1.0/2.0))
+
+            n = norm(at, 4)
+            assert_(issubclass(n.dtype.type, np.inexact))
+            assert_almost_equal(n, 2.0**(1.0/4.0))
+
+            n = norm(at, np.inf)
+            assert_(issubclass(n.dtype.type, np.inexact))
+            assert_almost_equal(n, 1.0)
+
+    def test_matrix_types(self):
+        a = np.array([[1, 0, 1], [0, 1, 1]])
+
+        exact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllInteger']
+        ]
+
+        inexact_types = [
+            np.dtype(_).type for _ in np.typecodes['AllFloat']
+        ]
+
+        all_types = exact_types + inexact_types
+
+        for each_inexact_types in all_types:
+            # These types are not supported by `linalg`.
+            if each_inexact_types in [np.float16, np.float128, np.complex256]:
```

Comment:
Turns out these aren't always defined. So, we need to remove these typecodes instead.